src/physics/nonlinear_solid.cpp

```diff
@@ -140,49 +140,159 @@ void NonlinearSolid::completeSetup()
     bc.project(displacement_);
   }
 
-  // The abstract mass bilinear form
-  std::unique_ptr<mfem::ParBilinearForm> M_form;
-
-  // The abstract viscosity bilinear form
-  std::unique_ptr<mfem::ParBilinearForm> S_form;
-
   // If dynamic, create the mass and viscosity forms
   if (timestepper_ != serac::TimestepMethod::QuasiStatic) {
     const double              ref_density = 1.0;  // density in the reference configuration
     mfem::ConstantCoefficient rho0(ref_density);
 
-    M_form = displacement_.createOnSpace<mfem::ParBilinearForm>();
+    M_ = displacement_.createOnSpace<mfem::ParBilinearForm>();
+    M_->AddDomainIntegrator(new mfem::VectorMassIntegrator(rho0));
+    M_->Assemble(0);
+    M_->Finalize(0);
+
+    M_mat_.reset(M_->ParallelAssemble());
 
-    M_form->AddDomainIntegrator(new mfem::VectorMassIntegrator(rho0));
-    M_form->Assemble(0);
-    M_form->Finalize(0);
+    C_ = displacement_.createOnSpace<mfem::ParBilinearForm>();
+    C_->AddDomainIntegrator(new mfem::VectorDiffusionIntegrator(*viscosity_));
+    C_->Assemble(0);
+    C_->Finalize(0);
 
-    S_form = displacement_.createOnSpace<mfem::ParBilinearForm>();
-    S_form->AddDomainIntegrator(new mfem::VectorDiffusionIntegrator(*viscosity_));
-    S_form->Assemble(0);
-    S_form->Finalize(0);
+    C_mat_.reset(C_->ParallelAssemble());
   }
 
-  // Set the MFEM abstract operators for use with the internal MFEM solvers
+  // We are assuming that the ODE is prescribing the
+  // acceleration value for the constrained dofs, so
+  // the residuals for those dofs can be taken to be zero.
+  //
+  // Setting iterative_mode to true ensures that these
+  // prescribed acceleration values are not modified by
+  // the nonlinear solve.
+  nonlin_solver_.nonlinearSolver().iterative_mode = true;
+  nonlin_solver_.SetOperator(residual_);
+
   if (timestepper_ == serac::TimestepMethod::QuasiStatic) {
-    nonlin_solver_.nonlinearSolver().iterative_mode = true;
-    nonlinear_oper_                                 = buildQuasistaticOperator(std::move(H_form));
-    nonlin_solver_.SetOperator(*nonlinear_oper_);
+    // the quasistatic case is entirely described by the residual,
+    // there is no ordinary differential equation
+    residual_ = StdFunctionOperator(
+        displacement_.space().TrueVSize(),
+
+        // residual function
+        [this](const mfem::Vector& u, mfem::Vector& r) mutable {
+          H_->Mult(u, r);  // r := H(u)
```

Comment:
From what I can tell `H` was promoted to a member variable so these functions wouldn't have a dangling reference later.

For a later PR, it might be worth refining the ownership model here - it looks like the `StdFunctionOperator` can take ownership of `H` but both its functions need to use it so it can't be moved into either lambda.

Reply:
I'm leaning toward keeping `H` a member variable of the physics module. Off the top of my head, design/topology optimization, eigenvalue problems, and custom preconditioners may want to know something about `H`, and it might get trickier to access if we narrow its scope any more.

Reply:
I agree. There are a lot of times when something other than the operator may want stiffness information.